api/lql.go

```diff
@@ -18,24 +18,201 @@
 
 package api
 
+import (
+	"encoding/json"
+	"net/url"
+	"regexp"
+	"strconv"
+	"time"
+
+	"github.com/pkg/errors"
+)
+
+const (
+	reLQL                  string = `(?ms)^(\w+)\([^)]+\)\s*{`
+	LQLQueryTranslateError string = "unable to translate query blob"
+)
+
+type LQLQuery struct {
+	ID             string `json:"LQL_ID,omitempty"`
+	StartTimeRange string `json:"START_TIME_RANGE,omitempty"`
+	EndTimeRange   string `json:"END_TIME_RANGE,omitempty"`
+	QueryText      string `json:"QUERY_TEXT"`
+	// QueryBlob is a special string that supports type conversion
+	// back and forth from LQL to JSON
+	QueryBlob string `json:"-"`
+}
+
+func (q *LQLQuery) Validate(allowEmptyTimes bool) error {
+	// translate
+	if err := q.Translate(); err != nil {
+		return err
+	}
+	// validate range
+	if err := q.ValidateRange(allowEmptyTimes); err != nil {
+		return err
+	}
+	// validate query
+	if q.QueryText == "" {
+		return errors.New("query should not be empty")
+	}
+	return nil
+}
+
+func (q *LQLQuery) Translate() error {
+	// query
+	if err := q.TranslateQuery(); err != nil {
+		return err
+	}
+	// start
+	if start, err := q.TranslateTime(q.StartTimeRange); err != nil {
+		return err
+	} else {
+		q.StartTimeRange = start
+	}
+	// end
+	if end, err := q.TranslateTime(q.EndTimeRange); err != nil {
+		return nil
+	} else {
+		q.EndTimeRange = end
+	}
+	return nil
+}
+
+func (q *LQLQuery) TranslateQuery() error {
+	// empty
+	if q.QueryText != "" {
+		return nil
+	}
+	// json
+	var t LQLQuery
+
+	if err := json.Unmarshal([]byte(q.QueryBlob), &t); err == nil {
+		if q.StartTimeRange == "" {
+			q.StartTimeRange = t.StartTimeRange
+		}
+		if q.EndTimeRange == "" {
+			q.EndTimeRange = t.EndTimeRange
+		}
+		q.QueryText = t.QueryText
+		return err
+	}
+	// lql
+	if matched, _ := regexp.MatchString(reLQL, q.QueryBlob); matched {
+		q.QueryText = q.QueryBlob
+		return nil
+	}
+	return errors.New(LQLQueryTranslateError)
+}
+
+func (q LQLQuery) TranslateTime(inTime string) (outTime string, err error) {
+	// empty
+	if inTime == "" {
+		return
+	}
+	// parse time as RFC3339
+	if t, err := time.Parse(time.RFC3339, inTime); err == nil {
+		outTime = t.UTC().Format(time.RFC3339)
+		return outTime, err
```

Comment:
when using named returns do not pass fields to the `return` statement
```suggestion
		return
```

Reply:
I thought named returns weren't developer friendly because you have to track what you're returning.  This seems like the best of both worlds since it makes it more obvious.  At this point the named returns are just a means of initializing the return variables...

Reply:
If this is something that we don't want to do, I can address this via the entire PR we don't need to comment each one.

Reply:
Fixed.